### DISCUSSION TITLE
Why was `onSuccess` in `useWaitForTransaction` deprecated?

### DISCUSSION THREAD

**Discussion by frankfka** [wevm/wagmi#2682] [General]
In `wagmi` 1.x, I see that `onSuccess`, etc. are now deprecated parameters for `useWaitForTransaction`. This was useful for triggering an effect (ex. invalidating queries) when a transaction succeeds. What was the rationale for removing this?

**Comment by jxom** [comment]
Can you not use the result in a `useEffect`?

**Reply by frankfka** [reply]
> Can you not use the result in a `useEffect`?

I can, yeah. Just had the convention of success / error callbacks, so was wondering if there's a specific reason for the deprecation

**Reply by jxom** [reply]
mainly the reasons here: https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose#a-bad-api

**Comment by generatorpoint** [comment]
Where is it mentioned that the `onSuccess` etc callbacks will be deprecated? I don't see it in the announcement or docs anywhere. Will this apply to the same callback options in `useWriteContract` @jxom ?

**Reply by tmm** [reply]
Mutations (e.g. `useContractWrite`, `useSendTransaction`, etc.) will continue to have callbacks, queries (e.g. `useContractRead`, `useWaitForTransaction`, etc.) will not. See https://github.com/wagmi-dev/wagmi/discussions/2682#discussioncomment-6349261 for more info.

**Reply by gpxl-dev** [reply]
So is this incorrect?
![image](https://github.com/wagmi-dev/wagmi/assets/82473383/4e6a2305-a32c-46f3-a817-1b98a350e596)

**Reply by tmm** [reply]
`usePrepareContractWrite` is a query, not a mutation, so yes, callbacks will be removed from that.

**Comment by gpxl-dev** [comment]
So there’s no way to callback after success of a prepared contract write?On 25 Aug 2023, at 22:22, awkweb ***@***.***> wrote:﻿
usePrepareContractWrite is a query, not a mutation, so yes, callbacks will be removed from that.

**Reply by tmm** [reply]
Check out https://tkdodo.eu/blog/breaking-react-querys-api-on-purpose for what to do instead.
